crates/dekoder/src/eko.rs

```diff
@@ -0,0 +1,140 @@
+//! Utilities for reading and writing an eko output.
+use std::collections::HashMap;
+use std::fs::remove_dir_all;
+use std::fs::File;
+use std::io::BufWriter;
+use std::path::PathBuf;
+use yaml_rust2::Yaml;
+
+use crate::{EKOError, Operator, Result};
+
+/// Default rel. error for the float comparison inside `EvolutionPoint`.
+const EP_CMP_RTOL: f64 = 1e-5;
+/// Default abs. error for the float comparison inside `EvolutionPoint`.
+const EP_CMP_ATOL: f64 = 1e-3;
+
+/// A reference point in the evolution atlas.
+pub struct EvolutionPoint {
+    /// Evolution scale.
+    pub scale: f64,
+    /// Number of flavors
+    pub nf: i64,
+}
+
+impl TryFrom<&Yaml> for EvolutionPoint {
+    type Error = EKOError;
+
+    /// Load from yaml.
+    fn try_from(yml: &Yaml) -> Result<Self> {
+        // work around float representation
+        let scale = yml["scale"].as_f64();
+        let scale = if scale.is_some() {
+            scale.ok_or(EKOError::KeyError(
+                "because failed to read scale as float".to_owned(),
+            ))?
+        } else {
+            yml["scale"].as_i64().ok_or(EKOError::KeyError(
+                "because failed to read scale as float from int".to_owned(),
+            ))? as f64
+        };
```

Comment:
Does this make sense? In the branch `if scale.is_some()` the variable `scale` is always `Some(...)` and thus `ok_or` never fails. In the else branch you're trying again with an integer, so it seems to me you want

```suggestion
        let scale = scale.map_or_else(
            || {
                let int = yml["scale"].as_i64().ok_or(EKOError::KeyError(
                    "because failed to read scale as float from int".to_owned(),
                ))?;
                Ok(i16::try_from(int)
                    // UNWRAP: integers unrepresentable as floating-point numbers are wrong inputs
                    .unwrap()
                    .into())
            },
            Ok::<_, EKOError>,
        )?;
```

Reply:
Sorry for merging too earlier. Yes, the `is_some` and `ok_or` was redundant, but this was my fault, because I wanted to ban all `unwrap` :upside_down_face: Your code works the same, but to me it looks more complicated ... would it be ok to just replace the first `ok_or` with `unwrap`? I'm also concerned about the integer range: `i16` might be too small (we need to read $Q^2$) - and I hope `as` does the same job as `into` ...

Reply:
You shouldn't ban `unwrap`, it's the right tool for [unrecoverable errors](https://doc.rust-lang.org/book/ch09-03-to-panic-or-not-to-panic.html). As for the integer case: 1) have you verified that this is necessary? I agree that `i16::MAX` is probably too close to the representable values that we're interested in. 2) `as` doesn't do the same as `into`, the former is approximate, the latter is exact. I believe we don't want approximate conversions.

Reply:
> and I hope `as` does the same job as `into` ...

It's slightly different:

> `as` is most commonly used to turn primitive types into other primitive types, but it has other uses that include turning pointers into addresses, addresses into pointers, and pointers into other pointers.
> https://doc.rust-lang.org/std/keyword.as.html

is more for primitives (as `f64`). While `.into()` is a method of a structured object, even represented by a trait (`Into`, paired with `From`).

> `as` can be seen as the primitive for `From` and `Into`: `as` only works with primitives (`u8`, `bool`, `str`, pointers, …) whereas `From` and `Into` also works with types like `String` or `Vec`.

(same page as above)

So, it is correct to use `as` for casting to `f64` (taking into account all the limitations of casting)

Reply:
@alecandido Either case is better than my solution. However, I'm not sure if reading integers is really necessary. The `... as f64` is possibly problematic as it's approximate.

Reply:
I was not aware of the lossless conversion with `.into()`. However, notice that it is not implemented from `i64` -> `f64` (and it makes sense, you definitely need more byte to represent an integer as a float without losing anything).
<img width="154" alt="image" src="https://github.com/user-attachments/assets/d5182954-06b0-4672-86ac-ebf22ca1a79b">

Reply:
- I like @alecandido 's suggestion from above as it is clear and concise - done in c9bb4ac0
- as can be seen in the list above there is no `From<i64>` so we need `as`; and I'm willing to accept an approximate conversion (which is _always_ there for int->float except a finite number of cases) as we don't have any physical precision to that numerical accuracy
- the casting from int branch is definitely needed and in fact the only tested branch with my test EKO which I got from running the LHA benchmarks
- in fact, this exposes another problematic implementation detail of yaml in Python (as for disadvantages of yaml) : the number in the yaml file is really `10000` - note the missing `.` - despite being originally a float
- as for the int range, stupid me: actually the calculation is easy :upside_down_face: we surely have scales large then $Q = 2^8\text{GeV}$

Reply:
This integer branch could be avoided if you'd used `serde` instead of manually parsing using `yaml`. In that case the Rust structs have the full type information and most of the YAML problems disappear.

Reply:
yes, yes @alecandido was already saying the same above :see_no_evil: - [I will try to do it later](https://github.com/NNPDF/eko/pull/400#discussion_r1724843377)

Reply:
Oh I missed that part of the discussion, but then you know that you should do it :smile: .